### PR TITLE
Hw3

### PR DESCRIPTION
**Pull request by vladislav77777**
**Add completed tasks and unit tests**

In this pull-requist I successfully completed all the tasks specified in the homework. Each assignment was carefully considered and decided as required.

In addition, I have written an extensive set of unit tests to ensure that my decisions are correct and reliable. Unit tests cover all functions and methods, and they successfully pass in an automated, continuous integration environment.

Changes:
Added Tasks 1-8 and TaskTests 1-8

Tests:
Unit tests for each function and method are written.
All tests successfully pass and cover various cases of use.
Code coverage by tests is 87.78%.

### CODE REVIEW

File: src/main/java/edu/hw3/Task2.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    public static List<String> clusterize(String s) {
+        Stack<Integer> stack = new Stack<>();
+        List<String> result = new ArrayList<>();
+        Integer openValue;
+        for (int i = 0; i < s.length(); i++) {
+            char c = s.charAt(i);
+            if (c == '(') {
+                stack.push(i);
+
+            } else if (c == ')' && s.charAt(stack.peek()) == '(') {
```
Comment (CrazyTosser) [inline src/main/java/edu/hw3/Task2.java]: Решение классное, но падает на вводе `)(`. И кстати, кавычки можно не вставлять руками, в задании это обозначало что массив строк вернуть

File: src/main/java/edu/hw3/Task5.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            return new String[] {};
+        }
+
+        Comparator<String> nameComparator = (contact1, contact2) -> {
+            String regex = "\\s+";
+            String[] parts1 = contact1.split(regex);
+            String[] parts2 = contact2.split(regex);
+            String lastName1 = parts1.length > 1 ? parts1[parts1.length - 1] : parts1[0];
+            String lastName2 = parts2.length > 1 ? parts2[parts2.length - 1] : parts2[0];
+            return order.equals("ASC") ? lastName1.compareTo(lastName2) : lastName2.compareTo(lastName1);
```
Comment (CrazyTosser) [inline src/main/java/edu/hw3/Task5.java]: Я дико извиняюсь, но в задании сказано
```
если фамилия отсутствует, то нужно использовать имя
возвращать нужно не строки, а объекты
```

File: src/main/java/edu/hw3/Task6/Stock.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package edu.hw3.Task6;
+
+public class Stock {
```
Comment (CrazyTosser) [inline src/main/java/edu/hw3/Task6/Stock.java]: FIO: Вместо этого можно писать `public record Stock(String symbol, double price) {}`. Кода меньше, читать легче

File: src/main/java/edu/hw3/Task5.java (hunk 2 above)
Comment (vladislav77777) [reply src/main/java/edu/hw3/Task5.java]: ага, исправил

**Pull request merged**